### ISSUE TITLE
"Installation.iso" not found error on reboot? File exists, but on target D: partition.

### ISSUE THREAD

**Issue by Mugsy323** [hakuna-m/wubiuefi#100]
I tried to use WUBI for Ubu-18.04 trying to install it to an empty D: partition (selected D: as target drive), but on reboot (performed by WUBI), I get an error telling me it is **"unable to find \ubuntu\install\installation.iso"**. But I checked and the file is there (on the target D: drive.)

Did I forget to set something? Is this a bug? TIA

**Comment by hakuna-m** [comment]
In the past, some users got also this error message. see e.g. #80 

A possible problem is always that Wubi runs on Windows but Ubuntu does not run on Windows. So Wubi provides all target drives on Windows. But for the second part of installation the target drive must be also available on Ubuntu.  

So you can check with `blkid` on busybox prompt, if the drive is available on Ubuntu, too.

**Comment by Mugsy323** [comment]
Thanks for the reply, but I'm not sure how this helps me?

Can I simply copy the "installation.iso" to c:\ubuntu\install\" (my Windows drive) and WUBI will find it? This won't wipe my C: drive will it? I need a Dual Boot system (note, I can't use a USB boot drive. My tablet's bios doesn't support it.)

**Comment by hakuna-m** [comment]
> I'm not sure how this helps me?

`blkid` shows the supported drives. If D: drive is not displayed and C: drive is displayed, you can use C: drive instead of D: drive. Ubuntu installed by Wubi(uefi) can share drives with Windows. If a USB boot drive is displayed, you can try to install on USB drive. IMHO, nativedisk support for GRUB or rEFInd are possible solution to boot from USB drives.

If `blkid` shows nothing, your system is currently not supported by the Ubuntu ISO which is used by Wubiuefi.

So it helps you to decide which drive is supported.

**Comment by Mugsy323** [comment]
GTK for the future, but for now, my USB ports are totally disabled.

No external drives. No keyboard. :(

**Comment by hakuna-m** [comment]
> No keyboard. :(

i.e. you cannot type cammands like **blkid** because of a missing keyboard ?

BTW: Please add existing threads like [ubuntuforums thread](https://ubuntuforums.org/showthread.php?t=2389939) to your question,

**Comment by Mugsy323** [comment]
Long story short, I accidentally disabled my tablets' USB ports in BIOS and am hoping I can reset it back to its defaults using Linux: (https://forums.linuxmint.com/viewtopic.php?t=226363#p1194543)

So I need a way to install Linux _entirely from Windows_ where I have access to the touch screen for typing.

https://ubuntuforums.org/showthread.php?t=2389939

**Comment by Mugsy323** [comment]
Long story short, I accidentally disabled my tablets' USB ports in BIOS and am hoping I can reset it back to its defaults using Linux: (https://forums.linuxmint.com/viewtopic.php?t=226363#p1194543)

So I need a way to install Linux entirely from Windows where I have access to the touch screen for typing.

https://ubuntuforums.org/showthread.php?t=2389939

**Comment by hakuna-m** [comment]
So you need a boot to an ISO with a Linux live system which supports your touch screen.

IMHO, Wubiuefi can help you a little bit. Wubiuefi installs a working GRUB boot loader. It is only an one-time boot by default but in your case it seems too risky to change normal boot order to boot always GRUB.

There is no need to repeat installation for an one-time boot. There are programs like [EasyUEFI](https://www.easyuefi.com/index-us.html) which can set one-time boots. see ["How to set a One-time EFI Boot Entry?"](https://www.easyuefi.com/faq/en-US/Set-Unset-One-time-Boot-Entry.html)

So it should be possible to boot to GRUB menu. You can modify GRUB menu by changing **\ubuntu\install\boot\grub\grub.cfg** on Windows.

Currently, the first entry ("Normal mode") is selected with
`set default=0`
But be free to change the default boot entry. If you want the 8th menu entry ("Demo mode") change it to
`set default=7`

If Ubuntu ISO does not work on your system, there are also other ISOs which you can test.

Examples: https://github.com/thias/glim/tree/master/grub2

**Comment by Mugsy323** [comment]
Thanks for the reply.

I got _close_ using "Unetbootin v6.57" and following the following instructions:
https://askubuntu.com/a/642253

It added a Windows-based Boot Menu at startup that looks like the "Windows Troubleshooter" with touchscreen support prompting me to choose either "Windows 10" or "UNetbootIn"... the latter of which is supposed to load the Ubuntu iso image UNetbootIn extracted when it was run from Windows.

Unfortunately, the last step failed, telling me it "could not find the disk image" (extracted to c:\unetboot\").

If I can figure out how to fix that, it should load the LiveCD, from which I could conceivably execute the "Modprobe" command w/o doing a full install of Ubuntu.

**Comment by Mugsy323** [comment]
Slight correction. It's not the Ubu iso it can't find (well, maybe. Dunno) it's "the Windows Boot Loader" it can't find when using Unetbootin.

A Google search shows this is a known problem trying to use Unetbootin to boot something on a device with a UEFI bios, and the "fix" is to go into bios and switch to "legacy mode."

If I could do that, I wouldn't be having this problem.

**Comment by Mugsy323** [comment]
Hey, this looks like great info. I got as far as trying to edit the .conf file with notepad, but there is no such file. The directory contains:

dir of s:\efi\ubuntu\wubildr\
grubia32.efi
grubx64.efi
hakumatata_wubi.cer
mmx64.efi
shimx64.efi

That's all. no .conf file. :(

**Comment by Mugsy323** [comment]
I downloaded the version from your link.

You didn't mention this, but I removed UNetbootin when it didn't work. Do I need to put it back first?

**Comment by hakuna-m** [comment]
> You didn't mention this, but I removed UNetbootin when it didn't work. Do I need to put it back first?

No, UNetbootin has nothing to do with Wubiuefi and rEFInd. 

> I downloaded the version from your link. 

Hmmm, it looks like another version is still installed. Did you uninstall your previous version and then install the [17101r328](https://github.com/hakuna-m/wubiuefi/releases/tag/17101r328) with Ubuntu 17.10.1 ?

The folder s:\EFI\ubuntu\wubildr should be removed during uninstall.

**Comment by Mugsy323** [comment]
Thx for the reply.

The previous version of WUBI didn't work so I uninstalled it.

Here was my process:

I reduced my installed programs & files (off-loading as much as I could to the cloud) to shrink the used space on my C: to free up as much space on the internal 32gb SSD as possible. I then created a 7.2GB D: partition (the minimum needed when I was trying to use UNetbootin.) Even after removing UNetbootin, I left the D: partition but reformatted it.

I then downloaded the r328 version of WUBI at your link and installed it.

I followed the WUBI installer steps and from an elevated CLI window, mounted an S: drive and tried to edit s:\EFI\ubuntu\wubildr\refind.conf using notepad.exe, but it said the file does not exist. So I checked the folder at that path and that's all I found.

**Comment by Mugsy323** [comment]
Quick question...

Should I even create a D: partition before installing WUBI? Without one, it wants to install to C: (and the last thing I need now is to screw up Windows.)

TIA

**Comment by hakuna-m** [comment]
The advantage of C: drive is that there is no need to create a partition which is also sometimes risky. If you install to C: drive it means that Wubi(uefi) creates the ubuntu folder on C: drive. It does not mean that it deletes all files on C: drive.

The advantage of D: drive is that there are independent file systems for Windows system drive C: and Ubuntu host drive D:. In some cases this provides more stability.

But it is up to you.

**Comment by Mugsy323** [comment]
My top concern at this point is not screwing up Windows b/c if I do there's no getting it back. It's the only means of doing anything at all left on this tablet.

What would help the most right now is just getting the "LiveCD" to load. I don't even need a *full* install, just access to the Ubu Linux terminal so I may issue the bios reset commands.

What would be the best method?

**Comment by hakuna-m** [comment]
> My top concern at this point is not screwing up Windows b/c if I do there's no getting it back

There are always remaining risks. The safest method for your Windows is to do nothing.  But the question is whether you can live without the missing UEFI BIOS functions.

> What would help the most right now is just getting the "LiveCD" to load.

If you use rEFInd, you can change the configuration file s:\EFI\ubuntu\wubildr\refind-wubi.conf to load the "LIveCD" only.

Just, copy the options line of the last submenuentry block ("Demo mode") to the first options line within main menuentry block.  The pattern of the first refind-wubi.conf file is [here](https://github.com/hakuna-m/wubiuefi/blob/refind/data/refind-install.cfg). Wubiuefi replaces the $ variables with values.

**Comment by Mugsy323** [comment]
Yeah, continuing without USB support isn't an option.

So, to be clear, installing WUBIUEFI to **C:** shouldn't harm Windows? What will the computer do when it starts? Show a boot menu?

Thx.

**Comment by hakuna-m** [comment]
> So, to be clear, installing WUBIUEFI to C: shouldn't harm Windows?

Yeah.

> What will the computer do when it starts? Show a boot menu?

A rEFInd boot menu like [this one](http://www.rodsbooks.com/refind).

There should be three selectable icons in the first row: An orange Ubuntu icon, a blue Windows icon and a penguin icon.

The rEFInd boot menu is defined by \EFI\ubuntu\wubildr\refind.conf on EFI partition (S: drive).

- The orange Ubuntu icon boots into Ubuntu (without GRUB). The default boot option installs Ubuntu.The boot options can be changed in \EFI\ubuntu\wubildr\refind-wubi.conf on EFI partition (S: drive).
- The Windows icon boots into Windows.
- The penguin icon offers a GRUB boot. The default boot option installs Ubuntu. GRUB boot can be changed in \ubuntu\install\boot\grub\grub.cfg on installation drive (C: or D: drive in your case).

**Comment by Mugsy323** [comment]
Okay, I removed the D: partition I created and followed your initial instructions on installing WUBIUEFI (including enabling "touch_screen" using notepad) to the C: drive.

On reboot, I got the menu but the first icon wasn't Ubu but instead an empty box. After a few seconds of not choosing an option, it (I assume) auto-selected it, which didn't work (but DOES appear to get me the UEFI command line, which would be useful if I had a keyboard. <LOL>

I powered off, rebooted, and the tablet went directly back into Windows (skipping the boot menu.)

I'm guessing it could not find the .iso. Wubi's installer appeared to install or download something, but it went way to quickly for a file that would have to be between 700MB and 1.5GB.

Did I do something wrong? Do I need to place the iso someplace in particular? Thx.

**Comment by hakuna-m** [comment]
> On reboot, I got the menu but the first icon wasn't Ubu but instead an empty box.

Was it the rEFInd boot menu ? Did Windows icon and penguin icon appear ?

I assume that C: drive has no drive name and volume name in S:\EFI\ubuntu\wubildr\refind-wubi.conf does not work.

Can you check the volume name in line 2 of S:\EFI\ubuntu\wubildr\refind-wubi.conf ?

**Comment by Mugsy323** [comment]
Yes, it was the rEFInd boot menu. The Windows & Penguin icons were there, but the first icon was a box with yellow square in it (IIRC).

The C: drive has no label. Windows is presently not reporting the existence of any other drives/partitions (no S:.)

I reinstalled WUBIUEFI and there is no "S:\EFI\ubuntu\wubildr\refind-wubi.conf" file.

OH CRAP! I reinstalled Wubi, edited "s:\EFI\ubuntu\wubildr\refind.conf" to uncomment "enable_touch", rebooted, no Ubuntu on the menu, but the touchpad doesn't work so I can't get back into Windows!

![menu](https://user-images.githubusercontent.com/33783094/39446271-8cae1ff0-4c83-11e8-9934-e5568b17b961.jpg)

Might I have accidentally uncommented some *other* "touch" setting and not noticed? Am I screwed?

**Comment by hakuna-m** [comment]
> The C: drive has no label.

Okay, I assumed that. If the volume name is empty, rEFInd finds only the parts on EFI partition (S:).

> but the touchpad doesn't work so I can't get back into Windows!

Some touchscreens work by uncommenting "enable_touch",  other touchscreens do not work. Maybe, the enabled touchscreen worked a little bit and stopped the timeout counter. 

But it should be still an one-time boot to rEFInd. So if you switch off your tablet, it should boot into Windows.

**Comment by hakuna-m** [comment]
But in your previous attempts, it was an one-time boot!?

Does it work, if you wait that rEFInd selects the first icon. Maybe an unsucessful boot, reboots to Windows.

**Comment by Mugsy323** [comment]
Yes, but in previous attempts, I had created a D: partition and set Wubi to install to it. :(

This time, I removed the D: partition and installed to C:.

**Comment by hakuna-m** [comment]
If I don't press any key in rEFInd menu, a message "Automatic boot in 10 seconds" appears. Then it counts down. In your picture there is no such a countdown:
![menu](https://user-images.githubusercontent.com/33783094/39446271-8cae1ff0-4c83-11e8-9934-e5568b17b961.jpg)
Maybe touchscreen or another key does work ?

> Yes, but in previous attempts, I had created a D: partition and set Wubi to install to it. :(

The EFI boot menu entry loads always EFI\ubuntu\wubildr\shimx64.efi on EFI partition (S:).  So there is no difference for EFI boot menu entry.

**Comment by Mugsy323** [comment]
There are only three buttons on the tablet: Volume Up/Down and Power. The touch screen doesn't work. Period.

And just as in Bios, pressing Volume down moves the selection down (to the "info" icon) and Volume Up moves it back. They only move in one direction. Power then selects whatever is highlighted.

Face it dude, you screwed me. :(

In future versions of WUBIUEFI, moving the cursor up/down should cycle through ALL the icons, not just the first row. And after repeated failures, Wubi should auto-remove itself. :(

**Comment by hakuna-m** [comment]
> In future versions of WUBIUEFI, moving the cursor up/down should cycle through ALL the icons, not just the first row.

It is an issue of rEFInd. There is no change for Wubiuefi. Maybe the [forum for rEFInd](https://sourceforge.net/p/refind/discussion/general) can help. Sorry, I didn't know that issue before. 

The standard version of Wubiuefi does not use rEFInd but some Apollo Lake devices do not work with GRUB (#84). So we created one version which contains Wubiuefi and rEFInd like existing customized ISOs for Apollo Lake devices.

> And after repeated failures, Wubi should auto-remove itself. :(

Wubiuefi creates an EFI boot menu entry with bcdedit option /addlast. i.e It should be always the last menu entry and the boot entry for Windows remains the first boot entry.

But if you insert a new entry, an one-time boot is the default value for most UEFI BIOS versions. Unfortunately, there are different  UEFI BIOS versions. There are UEFI BIOS versions which do not set an one-time boot. Other UEFI BIOS versions changes the one-time boot to a permanant boot. 

But in your case the one-time boot had worked. So I don't expected that issue.

**Comment by hakuna-m** [comment]
> not the rEFInd boot screen.

The configuration file **refind.conf** is for rEFInd only. For testing purpose, I enabled mouse support on my notebook (it has no touchscreen) by uncommenting "enable_mouse". Since then mouse selection in graphical rEFInd menu has been activated. It has never affected Ubuntu. In Ubuntu mouse support works fine even without using rEFInd.

Regarding comments in refind.conf, enable_touch feature is mutually exclusive with the enable_mouse feature. So in your case, it would be no good idea to enable mouse support. But did you enable mouse support ? 

> ```
> # Enable touch screen support. If active, this feature enables use of
> # touch screen controls (as on tablets). Note, however, that not all
> # tablets' EFIs provide the necessary underlying support, so this
> # feature may not work for you. If it does work, you should be able
> # to launch an OS or tool by touching it. In a submenu, touching
> # anywhere launches the currently-selection item; there is, at present,
> # no way to select a specific submenu item. This feature is mutually
> # exclusive with the enable_mouse feature. If both are uncommented,
> # the one read most recently takes precedence.
> ```

AFAIK, almost all UEFI devices provides a key sequence with internal keys (e.g. Volume up, Volume down, Power in your case) to enter UEFI firmware (BIOS) and UEFI boot menu. Wubiuefi does not delete Windows entry from UEFI boot menu. So it should be still possible to select it there.

**Comment by Mugsy323** [comment]
As noted previously, I can get into the Bios by pressing the Volume Up & Power buttons simultaneously during startup, but once I'm there, I can't access anything beyond the first menu. This is the entire reason I was trying to install Linux in hopes of resetting my Bios from it.

Windows is likely still fully intact on the internal drive, there is just no way of selecting it from the rEFInd menu using a touch screen (and the Volume/Power buttons only behave the same in rEFInd as they do in Bios.)

I did not enable the mouse or anything else in the .conf file other than the touch screen. Since the "mouse" in this case would be the touchpad on the keyboard, it would not have worked anyway.

As those hash-commented lines note: "not all tablets' EFIs provide the necessary underlying support". Looks like mine is one of them.

At this point, it appears my only hope left is to open the tablet up and attempt to wipe the Bios FlashRAM electronically. :(

**Comment by hakuna-m** [comment]
> I can get into the Bios by pressing the Volume Up & Power buttons simultaneously during startup, but once I'm there, I can't access anything beyond the first menu.

But maybe, there are two different things which you can still try:

- Does entering and leaving the UEFI BIOS reset the one-time-boot ?
- Is there another key sequence to enter UEFI boot menu ?

**Comment by Mugsy323** [comment]
> 
    Does entering and leaving the UEFI BIOS reset the one-time-boot ?
    Is there another key sequence to enter UEFI boot menu ?
>

No.
No.

**Comment by hakuna-m** [comment]
>> Is there another key sequence to enter UEFI boot menu ?

> No.

Do you have a manual which describes all key sequences for your tablet ? Or is the answer a result of your tries with different key sequences ?

**Comment by Mugsy323** [comment]
There are only 5 possible combinations:

Volume Down
Volume Up
Power
Volume Down + Power
Volume Up + Power

I've tried them all. Holding them down only cycles/repeats the keypress.

**Comment by hakuna-m** [comment]
> There are only 5 possible combinations:

Probably, there are no other combinations for pressing buttons simultaneously if Volume Up and Volume Down are not independent buttons. But there are also possible combinations which are not simultaneously. e.g. you can press Volume Up first and then Volume Down.

Some tablets have really tricky key sequences.

**Comment by Mugsy323** [comment]
The Vol Up/Down keys act exactly as arrow up/down. Pressing one then the other only moves the selections up & down.

The volume buttons are on the same rocker-arm.

This isn't like an Nintendo gamepad with "cheat code" sequence.

**Comment by hakuna-m** [comment]
> This isn't like a Nintendo gamepad with "cheat code" sequence.

You are right but e.g. key sequences for other tablets are ....

-    Press and hold Vol up button
-    Press and hold Power button for 5 seconds and keep holding the Vol up button
-    Release the Power button after 5 seconds but keep holding the Vol up button until you see BIOS menu

I don't know if you find such key sequences if you try just pressing Vol up + Power button simultaneously.

I also don't know if your tablet has such key sequences. But IMHO, it is not impossible.

**Comment by Mugsy323** [comment]
As I stated, I can already enter Bios by holding Volume Up & Power during boot, but once I am in there, I can not navigate to other pages because there is no way to move left/right.

All button presses have an immediate effect. The timing of keypresses have no more effect than holding down the arrow keys or Enter key on a regular keyboard.
